cmd/kubeadm/app/cmd/phases/reset/unmount_linux.go

```diff
@@ -45,8 +45,10 @@ func unmountKubeletDirectory(absoluteKubeletRunDirectory string) error {
 		if len(m) < 2 || !strings.HasPrefix(m[1], absoluteKubeletRunDirectory) {
 			continue
 		}
-		if err := syscall.Unmount(m[1], 0); err != nil {
+		if err := syscall.Unmount(m[1], syscall.MNT_DETACH); err != nil {
```

Comment:
before it was 0 without a flag.

would MNT_DETACH block until the unmount has finished and how much time can it take?

https://man7.org/linux/man-pages/man2/umount.2.html

Reply:
In both the ![docker](https://github.com/moby/moby/blob/97206b7cbadf97c0c6f447719e96189d023b33a7/libnetwork/osl/namespace_linux.go#L289) and [kata](https://github.com/kata-containers/kata-containers/blob/27dab249a0b434134ac0bee1595105d7ccbba363/src/runtime/virtcontainers/mount_linux.go#L114) containers, the default command is lazy unmount. The umount process will not get stuck, but the associated mount process (fuse) will wait until the process occupying the mount point is finished.

Reply:
I think this needs more discussion. `MNT_DETACH` allows the cleanup to continue safely, but it may delay the actual umount infinitely. For example, an admin may decide to detach the disk if he sees no mountpoint present, causing potential data damage. Do we have an example of doing lazy umount on persistent volume?

Both provided links are only related to bind mount. The docker one is mounting the network namespace, the kata one is container rootfs. I think they are very different from this case.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_helper_common.go#L44-L73 will do a force umount in kubelet side.